File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/repository/TransactionModelRowMapper.java
package com.example.tradelog.api.repository;

import com.example.common.converter.TimeConversion;
import com.example.tradelog.api.spec.model.TransactionModel;
import com.example.tradelog.api.spec.model.TransactionSettingsModel;
import com.example.tradelog.api.spec.model.TransactionType;

import java.sql.ResultSet;
import java.sql.SQLException;

class TransactionModelRowMapper {
    private ResultSet rs;

    TransactionModelRowMapper(ResultSet rs) {
        this.rs = rs;
    }

    TransactionModel invoke() throws SQLException {
        double preferredPrice = rs.getDouble("preferred_price");
        if (rs.wasNull()) {
            preferredPrice = 0.0;
        }

        TransactionSettingsModel optionsModel = new TransactionSettingsModel(
                rs.getString("id"),
                preferredPrice,
                rs.getBoolean("group_selected"),
                rs.getBoolean("leg_closed")
        );

        return new TransactionModel(
                rs.getString("id"),
                "",
                TimeConversion.fromTimestamp(rs.getTimestamp("date")),
                rs.getString("symbol"),
                TransactionType.Companion.lookup(rs.getString("transaction_type_fk")),
                rs.getDouble("broker_fees"),
                optionsModel);
    }
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/db/converter/TransactionModelRowMapper.kt
package com.example.tradelog.api.db.converter

import com.example.common.converter.TimeConverter
import com.example.tradelog.api.core.model.TransactionModel
import com.example.tradelog.api.core.model.TransactionSettingsModel
import com.example.tradelog.api.core.model.TransactionType
import java.sql.ResultSet
import java.util.*

class TransactionModelRowMapper(private var rs: ResultSet) {

    fun invoke(): TransactionModel {
        var preferredPrice = rs.getDouble("preferred_price")
        if (rs.wasNull()) {
           preferredPrice = 0.0
        }

        val settingsModel = TransactionSettingsModel(
                transactionId = rs.getObject("id", UUID::class.java),
                preferredPrice = preferredPrice,
                groupSelected = rs.getBoolean("group_selected"),
                legClosed = rs.getBoolean("leg_closed")
                )

        return TransactionModel(
                id = rs.getObject("id", UUID::class.java),
                accountId = rs.getObject("account_fk", UUID::class.java),
                portfolioId = rs.getObject("portfolio_fk", UUID::class.java),
                date = TimeConverter.fromTimestamp(rs.getTimestamp("date")),
                symbol = rs.getString("symbol"),
                type = TransactionType.lookup(rs.getString("transaction_type_fk")),
                brokerFees = rs.getDouble("broker_fees"),
                settings = settingsModel
        )
    }
}
